src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -112,9 +112,23 @@ public int Count
         
         public override int GetHashCode()
         {
-            return null == _array
-                        ? 0
-                        : _array.GetHashCode() ^ _offset ^ _count;
+            if (_array == null)
+            {
+                return 0;
+            }
+            
+#if !FEATURE_CORECLR
+            // Old implementation: Xor everything.
+            // This isn't optimal since it treats _offset and _count interchangeably,
+            // but we keep it around in the full framework for compat.
+            return _array.GetHashCode() ^ _offset ^ _count;
+#else
+            // Taken from ValueTuple.CombineHashCodes
+            int hash = _array.GetHashCode();
+            hash = ((hash << 5) + hash) ^ _offset;
+            hash = ((hash << 5) + hash) ^ _count;
```

Comment:
`(hash << 5) - hash` would be the equivalent of `hash * 31` which has been found to work relatively well as mult-and-xor (or mult-and-add) hashing goes (alas I can't find the literature I once read on this, it still produced much worse hashes than more complicated algorithms with strong avalanche effects, but they lose when it comes to calculation time). (Why it was found to work well is not clear; that it's a prime, that it's a Mersenne number and that both combine to mean its a Mersenne prime have each been offered as the reason, but it seems to do better than other numbers with the same features).